In [2]:
# Mathematical operations
import numpy as np

#Data Analysis - DE, DM, DC, DV
import pandas as pd

#dATA VISUALIZATION TOOL
import matplotlib.pyplot as plt

#Data Visaualization tool
import seaborn as sn

#serverless database
import sqlite3

#REGEX - DATA CLEANING
import re

# PLANNING/MY APPROACH

STEP1: connect database sqlit3.connect(database_name)
STEP2: Use the cursor function - Database_variable.cursor()


# WORKFLOW

1. I will establish a connection to the SQLITE Database by creating a connection object
2. I will create a cursor object using the cursor() method
3. I will execute the query cursor_variable.execute('query')
4. Fetch the data then use fetchall() method of the cursor variable/object
5. I will create a DataFrame w.r.t the SQlite
6. Data Analysis - Data Manipulation, Data Exploration, Data Cleaning, Data Visualization
7. I will write a conclusion at every step

Establishing a connection to the database

In [3]:
# Storing the path of the database into a variable 'db'
db = 'movies.sqlite'

#Creating a Database connection Object
conn = sqlite3.connect(db)

# database object that allows us to access data of one row at a time
# Creating a cursor object
cur = conn.cursor()

In [4]:
# Fetch all the data of the movies frm the movies from the movies table in the above database
cur.execute('SELECT * FROM movies')

#Creating a cursor object
movies = cur.fetchall()



In [31]:
# Displaying all the data from the movies table
#movies

In [5]:
# Creating a DataFrame
columns = ['id', 'original_title', 'budget', 'popularity','release_date', 'revenue','title','vote_average',
          'vote_count', 'overview', 'tagline', 'uid', 'director_id']
movies_df = pd.DataFrame(movies, columns= columns)
movies_df

id                            original_title     budget  popularity  \
0     43597                                    Avatar  237000000         150   
1     43598  Pirates of the Caribbean: At World's End  300000000         139   
2     43599                                   Spectre  245000000         107   
3     43600                     The Dark Knight Rises  250000000         112   
4     43601                               John Carter  260000000          43   
...     ...                                       ...        ...         ...   
4768  48395                               El Mariachi     220000          14   
4769  48396                                 Newlyweds       9000           0   
4770  48397                 Signed, Sealed, Delivered          0           1   
4771  48398                          Shanghai Calling          0           0   
4772  48399                         My Date with Drew          0           1   

     release_date     revenue                                     title  \
0      2009-12-10  2787965087                                    Avatar   
1      2007-05-19   961000000  Pirates of the Caribbean: At World's End   
2      2015-10-26   880674609                                   Spectre   
3      2012-07-16  1084939099                     The Dark Knight Rises   
4      2012-03-07   284139100                               John Carter   
...           ...         ...                                       ...   
4768   1992-09-04     2040920                               El Mariachi   
4769   2011-12-26           0                                 Newlyweds   
4770   2013-10-13           0                 Signed, Sealed, Delivered   
4771   2012-05-03           0                          Shanghai Calling   
4772   2005-08-05           0                         My Date with Drew   

      vote_average  vote_count  \
0              7.2       11800   
1              6.9        4500   
2              6.3        4466   
3              7.6        9106   
4              6.1        2124   
...            ...         ...   
4768           6.6         238   
4769           5.9           5   
4770           7.0           6   
4771           5.7           7   
4772           6.3          16   

                                               overview  \
0     In the 22nd century, a paraplegic Marine is di...   
1     Captain Barbossa, long believed to be dead, ha...   
2     A cryptic message from Bond’s past sends him o...   
3     Following the death of District Attorney Harve...   
4     John Carter is a war-weary, former military ca...   
...                                                 ...   
4768  El Mariachi just wants to play his guitar and ...   
4769  A newlywed couple's honeymoon is upended by th...   
4770  "Signed, Sealed, Delivered" introduces a dedic...   
4771  When ambitious New York attorney Sam is sent t...   
4772  Ever since the second grade when he first saw ...   

                                                tagline     uid  director_id  
0                           Enter the World of Pandora.   19995         4762  
1        At the end of the world, the adventure begins.     285         4763  
2                                 A Plan No One Escapes  206647         4764  
3                                       The Legend Ends   49026         4765  
4                  Lost in our world, found in another.   49529         4766  
...                                                 ...     ...          ...  
4768  He didn't come looking for trouble, but troubl...    9367         5097  
4769  A newlywed couple's honeymoon is upended by th...   72766         6485  
4770                                               None  231617         7108  
4771                           A New Yorker in Shanghai  126186         7109  
4772                                               None   25975         7110  

[4773 rows x 13 columns]

In [16]:
# Display the dataframe
movies_df

id                            original_title     budget  popularity  \
0     43597                                    Avatar  237000000         150   
1     43598  Pirates of the Caribbean: At World's End  300000000         139   
2     43599                                   Spectre  245000000         107   
3     43600                     The Dark Knight Rises  250000000         112   
4     43601                               John Carter  260000000          43   
...     ...                                       ...        ...         ...   
4768  48395                               El Mariachi     220000          14   
4769  48396                                 Newlyweds       9000           0   
4770  48397                 Signed, Sealed, Delivered          0           1   
4771  48398                          Shanghai Calling          0           0   
4772  48399                         My Date with Drew          0           1   

     release_date     revenue                                     title  \
0      2009-12-10  2787965087                                    Avatar   
1      2007-05-19   961000000  Pirates of the Caribbean: At World's End   
2      2015-10-26   880674609                                   Spectre   
3      2012-07-16  1084939099                     The Dark Knight Rises   
4      2012-03-07   284139100                               John Carter   
...           ...         ...                                       ...   
4768   1992-09-04     2040920                               El Mariachi   
4769   2011-12-26           0                                 Newlyweds   
4770   2013-10-13           0                 Signed, Sealed, Delivered   
4771   2012-05-03           0                          Shanghai Calling   
4772   2005-08-05           0                         My Date with Drew   

      vote_average  vote_count  \
0              7.2       11800   
1              6.9        4500   
2              6.3        4466   
3              7.6        9106   
4              6.1        2124   
...            ...         ...   
4768           6.6         238   
4769           5.9           5   
4770           7.0           6   
4771           5.7           7   
4772           6.3          16   

                                               overview  \
0     In the 22nd century, a paraplegic Marine is di...   
1     Captain Barbossa, long believed to be dead, ha...   
2     A cryptic message from Bond’s past sends him o...   
3     Following the death of District Attorney Harve...   
4     John Carter is a war-weary, former military ca...   
...                                                 ...   
4768  El Mariachi just wants to play his guitar and ...   
4769  A newlywed couple's honeymoon is upended by th...   
4770  "Signed, Sealed, Delivered" introduces a dedic...   
4771  When ambitious New York attorney Sam is sent t...   
4772  Ever since the second grade when he first saw ...   

                                                tagline     uid  director_id  
0                           Enter the World of Pandora.   19995         4762  
1        At the end of the world, the adventure begins.     285         4763  
2                                 A Plan No One Escapes  206647         4764  
3                                       The Legend Ends   49026         4765  
4                  Lost in our world, found in another.   49529         4766  
...                                                 ...     ...          ...  
4768  He didn't come looking for trouble, but troubl...    9367         5097  
4769  A newlywed couple's honeymoon is upended by th...   72766         6485  
4770                                               None  231617         7108  
4771                           A New Yorker in Shanghai  126186         7109  
4772                                               None   25975         7110  

[4773 rows x 13 columns]

In [18]:
movies_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4773 entries, 0 to 4772
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   id              4773 non-null   int64  
 1   original_title  4773 non-null   object 
 2   budget          4773 non-null   int64  
 3   popularity      4773 non-null   int64  
 4   release_date    4773 non-null   object 
 5   revenue         4773 non-null   int64  
 6   title           4773 non-null   object 
 7   vote_average    4773 non-null   float64
 8   vote_count      4773 non-null   int64  
 9   overview        4770 non-null   object 
 10  tagline         3951 non-null   object 
 11  uid             4773 non-null   int64  
 12  director_id     4773 non-null   int64  
dtypes: float64(1), int64(7), object(5)
memory usage: 484.9+ KB


In [32]:
# Fetch all the data of the movies frm the directors from the director table in the above database
cur.execute('SELECT * FROM directors')

#Creating a cursor object
directors = cur.fetchall()
#directors

In [8]:
# Creating a dataframe 
directors_df = pd.DataFrame(directors, columns=['name', 'director_id', 'gender','uid','department'])
directors_df

name  director_id  gender      uid department
0         James Cameron         4762       2     2710  Directing
1        Gore Verbinski         4763       2     1704  Directing
2            Sam Mendes         4764       2       39  Directing
3     Christopher Nolan         4765       2      525  Directing
4        Andrew Stanton         4766       2        7  Directing
...                 ...          ...     ...      ...        ...
2344      Shane Carruth         7106       2    76624  Directing
2345   Neill Dela Llana         7107       0  1174437  Directing
2346        Scott Smith         7108       0  1219158  Directing
2347        Daniel Hsia         7109       2   208138  Directing
2348   Brian Herzlinger         7110       2    85563  Directing

[2349 rows x 5 columns]

In [23]:
directors_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2349 entries, 0 to 2348
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   name         2349 non-null   object
 1   director_id  2349 non-null   int64 
 2   gender       2349 non-null   int64 
 3   uid          2349 non-null   int64 
 4   department   2349 non-null   object
dtypes: int64(3), object(2)
memory usage: 91.9+ KB


In [24]:
movies_df.shape

(4773, 13)

In [25]:
directors_df.shape

(2349, 5)

In [26]:
# Statistical view
movies_df.describe()

id        budget   popularity       revenue  vote_average  \
count   4773.000000  4.773000e+03  4773.000000  4.773000e+03   4773.000000   
mean   45985.680913  2.922461e+07    21.136602  8.277710e+07      6.109177   
std     1382.032760  4.078679e+07    31.864132  1.632374e+08      1.146789   
min    43597.000000  0.000000e+00     0.000000  0.000000e+00      0.000000   
25%    44790.000000  9.500000e+05     4.000000  0.000000e+00      5.600000   
50%    45983.000000  1.500000e+07    13.000000  1.947811e+07      6.200000   
75%    47176.000000  4.000000e+07    28.000000  9.361701e+07      6.800000   
max    48399.000000  3.800000e+08   875.000000  2.787965e+09     10.000000   

         vote_count            uid  director_id  
count   4773.000000    4773.000000  4773.000000  
mean     694.541588   56244.396396  5586.661429  
std     1237.251355   87452.149183   670.413263  
min        0.000000       5.000000  4762.000000  
25%       55.000000    8998.000000  5011.000000  
50%      238.000000   14517.000000  5389.000000  
75%      742.000000   57876.000000  6086.000000  
max    13752.000000  459488.000000  7110.000000

# QUESTION1: How many movies are present in the moovies table

In [27]:
movies_df.columns

Index(['id', 'original_title', 'budget', 'popularity', 'release_date',
       'revenue', 'title', 'vote_average', 'vote_count', 'overview', 'tagline',
       'uid', 'director_id'],
      dtype='object')

In [29]:
cur.execute('SELECT COUNT(id) FROM movies')
count = cur.fetchall()
count
print(f"The number of movies present in the movies table is {count[0][0]}")

The number of movies present in the movies table is 4773


# QUESTION2: Find out these 3 directors : James Cameron, Luc Besson, John Woo

In [30]:
directors_df.columns

Index(['name', 'director_id', 'gender', 'uid', 'department'], dtype='object')

In [31]:
cur.execute("SELECT * FROM directors WHERE name IN ('james Cameron','Luc Besson','John Woo')")
three_directors = cur.fetchall()
print(f"Here are the info about the three directors {three_directors}")

Here are the info about the three directors [('John Woo', 4893, 2, 11401, 'Directing'), ('Luc Besson', 4949, 2, 59, 'Directing')]


# QUESTION3: Find all the directors with the names starting with Steven

In [32]:
cur.execute("SELECT * FROM directors WHERE name LIKE 'Steven%'") #Like shows a pattern
name_like = cur.fetchall()
print(f"These are the directors starting with 'Steven' {name_like}")

These are the directors starting with 'Steven' [('Steven Spielberg', 4799, 2, 488, 'Directing'), ('Steven Soderbergh', 4909, 2, 1884, 'Directing'), ('Steven Brill', 5013, 2, 32593, 'Directing'), ('Steven Zaillian', 5117, 2, 2260, 'Directing'), ('Steven Quale', 5216, 2, 93214, 'Directing'), ('Steven Seagal', 5221, 2, 23880, 'Directing'), ('Steven E. de Souza', 5390, 2, 1726, 'Directing'), ('Steven Shainberg', 5803, 2, 67795, 'Directing'), ('Steven R. Monroe', 6713, 2, 88039, 'Directing')]


# QUESTION4: Count all the female directors Hint: Gender=1

In [35]:
cur.execute("SELECT COUNT(name) from directors where gender = 1")
female_directors = cur.fetchall()
print(f"The number of female directors is {female_directors[0][0]}")

The number of female directors is 150


There are 150 female directors from the directors table

# QUESTION5: Find the name of the 10th First Female directors

In [38]:
cur.execute("SELECT name from directors where gender = 1")
tenth_female = cur.fetchall()
print(f"The 10th female directors is {tenth_female[9][0]}")

The 10th female directors is Angelina Jolie


# QUESTION6: Find the top 3 most popular movies

In [39]:
movies_df.columns

Index(['id', 'original_title', 'budget', 'popularity', 'release_date',
       'revenue', 'title', 'vote_average', 'vote_count', 'overview', 'tagline',
       'uid', 'director_id'],
      dtype='object')

In [40]:
cur.execute("SELECT original_title from movies ORDER BY popularity desc LIMIT 3")
top_ten = cur.fetchall()
print(f"These are the top 10 most popular movies {top_ten}")

These are the top 10 most popular movies [('Minions',), ('Interstellar',), ('Deadpool',)]


# QUESTION7: Find the top three most bankable movies(huge budget)
   

In [42]:
movies_df.columns

Index(['id', 'original_title', 'budget', 'popularity', 'release_date',
       'revenue', 'title', 'vote_average', 'vote_count', 'overview', 'tagline',
       'uid', 'director_id'],
      dtype='object')

In [43]:
cur.execute("SELECT original_title FROM movies ORDER BY budget desc LIMIT 3")
most_bankable = cur.fetchall()
print(f"The top three most bankable movies are {most_bankable}")

The top three most bankable movies are [('Pirates of the Caribbean: On Stranger Tides',), ("Pirates of the Caribbean: At World's End",), ('Avengers: Age of Ultron',)]


# QUESTION8:  Find out the most Awarded avarage rated movie since January  1st, 2000

In [44]:
movies_df.columns

Index(['id', 'original_title', 'budget', 'popularity', 'release_date',
       'revenue', 'title', 'vote_average', 'vote_count', 'overview', 'tagline',
       'uid', 'director_id'],
      dtype='object')

In [10]:
cur.execute("SELECT original_title FROM movies WHERE release_date>'2000-01-01' ORDER BY vote_average DESC limit 1 ")
most_awarded = cur.fetchall()
print(f"The most awarded average rated movie is {most_awarded[0][0]}") 

The most awarded average rated movie is Sardaarji


# QUESTION9: Find all the movies that has been directed by Brenda Chapman

In [11]:
movies_df.columns

Index(['id', 'original_title', 'budget', 'popularity', 'release_date',
       'revenue', 'title', 'vote_average', 'vote_count', 'overview', 'tagline',
       'uid', 'director_id'],
      dtype='object')

In [12]:
directors_df.columns

Index(['name', 'director_id', 'gender', 'uid', 'department'], dtype='object')

In [13]:
query = """SELECT original_title FROM movies LEFT JOIN directors on movies.director_id = directors.id WHERE directors.name ='Brenda Chapman' """
cur.execute(query)
directed_movie = cur.fetchall()
print(f"The movies directed by Brenda Chapman is/are {directed_movie[0][0]}")

The movies directed by Brenda Chapman is/are Brave


# QUESTION10: Find out the director who has made the most movies

In [15]:
movies_df.columns

Index(['id', 'original_title', 'budget', 'popularity', 'release_date',
       'revenue', 'title', 'vote_average', 'vote_count', 'overview', 'tagline',
       'uid', 'director_id'],
      dtype='object')

In [16]:
directors_df.columns

Index(['name', 'director_id', 'gender', 'uid', 'department'], dtype='object')

In [21]:
cur.execute("SELECT name FROM directors JOIN movies ON directors.id = movies.director_id GROUP BY director_id ORDER BY COUNT(name) DESC LIMIT 1")
most_movies = cur.fetchall()
print(f"This is the director who made the most movies {most_movies[0][0]}")

This is the director who made the most movies Steven Spielberg


# QUESTION11: Find out the director who is most bankable

In [23]:
cur.execute('SELECT name FROM directors JOIN movies ON directors.id = movies.director_id  GROUP BY director_id ORDER BY SUM(budget) DESC LIMIT 1 ')
director_most_bankable = cur.fetchall()
print(f"The most bankable director is {director_most_bankable[0][0]}")

The most bankable director is Steven Spielberg


# QUESTION12: Find out the Top 10 highest budget making movies

In [24]:
cur.execute('SELECT original_title from movies order by budget desc limit 10')
top_10 = cur.fetchall()
print(f"{top_10}")

[('Pirates of the Caribbean: On Stranger Tides',), ("Pirates of the Caribbean: At World's End",), ('Avengers: Age of Ultron',), ('Superman Returns',), ('John Carter',), ('Tangled',), ('Spider-Man 3',), ('The Lone Ranger',), ('The Dark Knight Rises',), ('Harry Potter and the Half-Blood Prince',)]


# QUESTION13: Find the top 10 revenue making movies

In [26]:
cur.execute('SELECT original_title from movies order by revenue desc limit 10 ')
top_10_revenue = cur.fetchall()
print(f"The top 10 revenue making movies are {top_10_revenue}")

The top 10 revenue making movies are [('Avatar',), ('Titanic',), ('The Avengers',), ('Jurassic World',), ('Furious 7',), ('Avengers: Age of Ultron',), ('Frozen',), ('Iron Man 3',), ('Minions',), ('Captain America: Civil War',)]


# QUESTION14: Find out the top 10 most popular movies with the highest rating (vote_average)

In [27]:
movies_df.columns

Index(['id', 'original_title', 'budget', 'popularity', 'release_date',
       'revenue', 'title', 'vote_average', 'vote_count', 'overview', 'tagline',
       'uid', 'director_id'],
      dtype='object')

In [30]:
cur.execute('SELECT original_title from movies order by popularity,  vote_average desc limit 10')
top_10_rating = cur.fetchall()
print(f"The top 10 most popular movies w.r.t the highest ratings are {top_10_rating}")

The top 10 most popular movies w.r.t the highest ratings are [('Stiff Upper Lips',), ('Dancer, Texas Pop. 81',), ('Sardaarji',), ("One Man's Hero",), ('There Goes My Baby',), ('Counting',), ('Déjà Vu',), ('Dolphins and Whales: Tribes of the Ocean',), ('Lake of Fire',), ('Dreaming of Joseph Lees',)]


# QUESTION15: Find the top 10 directors with the number of movies and revenue where revenue should be taken into account for doing the new analysis     

The director who has got the highest revenue should comes at the top 

# QUESTION16: Find the top 10 directors with the number of movies and revenue where movies should be taken into account for doing the new analysis     

# QUESTION17: Give the title, release date, budget, revenue, popularity and vote average of those movies made by steven Spielberg